File: Embarcadero/Berlin/Object Pascal/TestBed/MainForm.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit MainForm;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants,
  FMX.Types, FMX.Controls, FMX.Forms, FMX.Graphics, FMX.Dialogs, FMX.StdCtrls, FMX.Objects, FMX.ListBox,
  FMX.Layouts, FMX.Controls.Presentation,
  Box2D.Collision, Box2D.Common, Box2D.Dynamics, Box2D.Rope, Box2DTypes, Test;

type
  TTestBedForm = class(TForm)
    MainPanel: TPanel;
    RightPanel: TPanel;
    RightLayout: TLayout;
    TestLabel: TLabel;
    TestList: TComboBox;
    Line1: TLine;
    EnableLayout: TLayout;
    Sleep: TCheckBox;
    WarmStarting: TCheckBox;
    TimeOfImpact: TCheckBox;
    SubStepping: TCheckBox;
    DrawLayout: TLayout;
    ShapesChk: TCheckBox;
    AABBsChk: TCheckBox;
    ContactPointsChk: TCheckBox;
    ContactNormalsChk: TCheckBox;
    ContactImpulsesChk: TCheckBox;
    FrictionImpulsesChk: TCheckBox;
    CenterOfMassesChk: TCheckBox;
    ProfileChk: TCheckBox;
    StatisticsChk: TCheckBox;
    JointsChk: TCheckBox;
    ButtonsLayout: TLayout;
    QuiBtn: TButton;
    RestartBtn: TButton;
    SingleStepBtn: TButton;
    PauseBtn: TButton;
    PaintBox: TPaintBox;
    RightSplitter: TSplitter;
    Timer1: TTimer;
    procedure FormShow(Sender: TObject);
    procedure FormResize(Sender: TObject);
    procedure Timer1Timer(Sender: TObject);
    procedure PaintBoxPaint(Sender: TObject; Canvas: TCanvas);
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure TestListChange(Sender: TObject);
    procedure FormKeyDown(Sender: TObject; var Key: Word; var KeyChar: Char; Shift: TShiftState);
    procedure FormKeyUp(Sender: TObject; var Key: Word; var KeyChar: Char; Shift: TShiftState);
    procedure PaintBoxMouseDown(Sender: TObject; Button: TMouseButton; Shift: TShiftState; X, Y: Single);
    procedure PaintBoxMouseMove(Sender: TObject; Shift: TShiftState; X, Y: Single);
    procedure PaintBoxMouseUp(Sender: TObject; Button: TMouseButton; Shift: TShiftState; X, Y: Single);
    procedure PauseBtnClick(Sender: TObject);
    procedure SingleStepBtnClick(Sender: TObject);
    procedure RestartBtnClick(Sender: TObject);
    procedure QuiBtnClick(Sender: TObject);
    procedure DrawLayoutChanged(Sender: TObject);
  private
    { Private declarations }
    testIndex: Int32;
    testSelection: Int32;
    testCount: Int32;
    entry: PTestEntry;
    test: TTest;
    settings: Settings;
    rightMouseDown: Boolean;
    lastp: b2Vec2;
  public
    { Public declarations }
    procedure LoadTests;
    procedure TestChanged;
    procedure ResetView;
  end;

var
  TestBedForm: TTestBedForm;

implementation

{$R *.fmx}

uses
  System.Character,
  DebugDraw;

procedure TTestBedForm.DrawLayoutChanged(Sender: TObject);
begin
  settings.drawShapes := ShapesChk.IsChecked;
  settings.drawJoints := JointsChk.IsChecked;
  settings.drawAABBs := AABBsChk.IsChecked;
  settings.drawContactPoints := ContactPointsChk.IsChecked;
  settings.drawContactNormals := ContactNormalsChk.IsChecked;
  settings.drawContactImpulse := ContactImpulsesChk.IsChecked;
  settings.drawFrictionImpulse := FrictionImpulsesChk.IsChecked;
  settings.drawCOMs := CenterOfMassesChk.IsChecked;
  settings.drawStats := StatisticsChk.IsChecked;
  settings.drawProfile := ProfileChk.IsChecked;
end;

procedure TTestBedForm.FormCreate(Sender: TObject);
begin
  settings := Settings.Create;
  LoadTests;

end;

procedure TTestBedForm.FormDestroy(Sender: TObject);
begin
  Timer1.Enabled := False;
  test.Free;
  test := nil;
end;

procedure TTestBedForm.FormKeyDown(Sender: TObject; var Key: Word; var KeyChar: Char; Shift: TShiftState);
var
  modCtrl: Boolean;
  newOrigin: b2Vec2;
begin
  modCtrl := ssCtrl in Shift;
  case Key of
    vkEscape: Close;

    vkLeft:
    begin
      if modCtrl and (test <> nil) then
      begin
        newOrigin := b2vec2.Create(2.0, 0.0);
        test.ShiftOrigin(newOrigin);
      end
      else
      begin
        g_camera.m_center.x  :=  g_camera.m_center.x - 1.0;
      end;

    end;

    vkRight:
      if modCtrl and (test <> nil) then
      begin
        newOrigin := b2vec2.Create(-2.0, 0.0);
        test.ShiftOrigin(newOrigin);
      end
      else
      begin
        g_camera.m_center.x := g_camera.m_center.x + 1.0;
      end;

    vkDown:
      if modCtrl and (test <> nil) then
      begin
        newOrigin := b2vec2.Create(0.0, 2.0);
        test.ShiftOrigin(newOrigin);
      end
      else
      begin
        g_camera.m_center.y := g_camera.m_center.y - 1.0;
      end;

    vkUp:
      if modCtrl and (test <> nil) then
      begin
        newOrigin := b2vec2.Create(0.0, -2.0);
        test.ShiftOrigin(newOrigin);
      end
      else
      begin
        g_camera.m_center.y := g_camera.m_center.y + 1.0;
      end;

    vkHome:
    begin
      // Reset view
      g_camera.m_zoom := 1.0;
      g_camera.m_center.&Set(0.0, 20.0);
    end;

    vkSpace:
      if test <> nil then
      begin
        test.LaunchBomb;
      end;
  else
    case KeyChar of
      'P', 'p':
        settings.pause := not settings.pause;

      ' ':
        if test <> nil then
        begin
          test.LaunchBomb;
        end;

      'R', 'r':
        begin
          test.Free;
          test := entry^.createFcn;
        end;
    else
      if test <> nil then
      begin
        test.Keyboard(Ord(KeyChar.ToUpper));
      end;
    end;
  end;

end;

procedure TTestBedForm.FormKeyUp(Sender: TObject; var Key: Word; var KeyChar: Char; Shift: TShiftState);
begin
  if test <> nil then
  begin
    test.KeyboardUp(Ord(KeyChar.ToUpper));
  end;
end;

procedure TTestBedForm.FormResize(Sender: TObject);
begin
  ResetView;
end;

procedure TTestBedForm.FormShow(Sender: TObject);
begin
  ResetView;
  DrawLayoutChanged(nil);
  TestChanged;
end;

procedure TTestBedForm.LoadTests;
var
  entry: PTestEntry;
begin
  testCount := 0;
  while (testCount < Length(g_testEntries)) do
  begin
    entry := @g_testEntries[testCount];
    TestList.Items.Add(entry^.name);
    Inc(testCount);
  end;

  TestList.ItemIndex := 0;
  testSelection := 0;
  testIndex := -1;
  test := nil;
end;

procedure TTestBedForm.PaintBoxMouseDown(Sender: TObject; Button: TMouseButton; Shift: TShiftState; X, Y: Single);
var
  pt: b2Vec2;
begin
  if (Button = TMouseButton.mbLeft) and (test <> nil) then
  begin
    pt := g_debugDraw.ScreenToWorld(X, Y);
    if ssShift in Shift then
    begin
      test.ShiftMouseDown(pt);
    end
    else
    begin
      test.MouseDown(pt);
    end;
  end;
end;

procedure TTestBedForm.PaintBoxMouseMove(Sender: TObject; Shift: TShiftState; X, Y: Single);
var
  pt: b2Vec2;
begin
  if test <> nil then
  begin
    pt := g_debugDraw.ScreenToWorld(X, Y);
    test.MouseMove(pt);
  end;
end;

procedure TTestBedForm.PaintBoxMouseUp(Sender: TObject; Button: TMouseButton; Shift: TShiftState; X, Y: Single);
var
  pt: b2Vec2;
begin
  if test <> nil then
  begin
    pt := g_debugDraw.ScreenToWorld(X, Y);
    test.MouseUp(pt);
  end;

end;

procedure TTestBedForm.PaintBoxPaint(Sender: TObject; Canvas: TCanvas);
var
  rect: TRectF;
begin
  Canvas.BeginScene;
  try
    Canvas.Fill.Color := b2Color2TColor(b2Color.Create(0.3, 0.3, 0.3, 1));
    rect := PaintBox.BoundsRect;
    Canvas.FillRect(rect, 0, 0, [], DEFAULT_OPACITY);

    if test <> nil then

    begin

      g_debugDraw.Canvas := Canvas;

      g_debugDraw.Canvas.Fill.Color := TAlphaColorRec.Yellow;

      test.DrawTitle(entry.name);

      g_debugDraw.Canvas.Fill.Color := TAlphaColorRec.Aqua;

      test.Step(@settings);

    end;

  finally
    Canvas.EndScene;
  end;
end;

procedure TTestBedForm.PauseBtnClick(Sender: TObject);
begin
  settings.pause := not settings.pause;
end;

procedure TTestBedForm.QuiBtnClick(Sender: TObject);
begin
  Close;
end;

procedure TTestBedForm.ResetView;
begin
  g_debugDraw.CanvasHeight := PaintBox.Height;
  g_camera.m_width := Round(PaintBox.Width);
  g_camera.m_height:= Round(PaintBox.Height);
  g_camera.m_center.&Set(PaintBox.Width/2, PaintBox.Height/2);
end;

procedure TTestBedForm.RestartBtnClick(Sender: TObject);
begin
  test.Free;
  test := entry.createFcn;
end;

procedure TTestBedForm.SingleStepBtnClick(Sender: TObject);
begin
  settings.singleStep := not settings.singleStep;
end;

procedure TTestBedForm.TestChanged;
begin
  if (testSelection <> testIndex) then
  begin
    testIndex := testSelection;
    test.Free;
    entry := @g_testEntries[testIndex];
    test := entry^.createFcn;
  end;

end;

procedure TTestBedForm.TestListChange(Sender: TObject);
begin
  if (TestList.ItemIndex <> testSelection) then
  begin
    testSelection := TestList.ItemIndex;
    TestChanged;
  end;
end;

procedure TTestBedForm.Timer1Timer(Sender: TObject);
begin
  if test <> nil then
    Invalidate;
end;


end.

File: Embarcadero/Berlin/CPP/TestBed/MainForm.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

#include <fmx.h>
#pragma hdrstop

#include "MainForm.h"
#include "DebugDraw.h"
#include "Test.h"
#include "DrawUtil.h"
#include <System.Character.hpp>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"
TTestBedForm *TestBedForm;


namespace {
  int32 testIndex = 0;
  int32 testSelection = 0;
  int32 testCount = 0;
  TestEntry* entry;
  Test* test;
  Settings settings;
  bool rightMouseDown;
  b2Vec2 lastp;
}

//---------------------------------------------------------------------------
__fastcall TTestBedForm::TTestBedForm(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TTestBedForm::FormCreate(TObject *Sender)
{
  LoadTests();
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::FormDestroy(TObject *Sender)
{
  Timer1->Enabled = false;
  Test* t = test;
  test = 0;
  delete t;
}
//---------------------------------------------------------------------------

void TTestBedForm::LoadTests()
{
  testCount = 0;
  while (g_testEntries[testCount].createFcn != NULL)
  {
    TestEntry& entry = g_testEntries[testCount];
    TestList->Items->Add(entry.name);
    ++testCount;
  }

  TestList->ItemIndex = 0;
  testSelection = 0;
  testIndex = -1;
  test = 0;
}
//---------------------------------------------------------------------------

void TTestBedForm::TestChanged()
{
  if (testSelection != testIndex)
  {
    testIndex = testSelection;
    delete test;
    entry = g_testEntries + testIndex;
    test = entry->createFcn();
  }
}
//---------------------------------------------------------------------------

void TTestBedForm::ResetView()
{
  g_debugDraw.ScaleX = DEFAULT_SCALE;
  g_debugDraw.ScaleY = DEFAULT_SCALE;
  g_debugDraw.CanvasHeight = PaintBox->Height;
  g_camera.m_width = PaintBox->Width;
  g_camera.m_height = PaintBox->Height;
  g_camera.m_center.Set(PaintBox->Width/2, PaintBox->Height/2);
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::PaintBoxPaint(TObject *Sender, TCanvas *Canvas)
{
  Canvas->BeginScene();
  __try
  {
	Canvas->Fill->Color = b2Color2TColor(b2Color(0.3, 0.3, 0.3, 1));
	TRectF rect = PaintBox->BoundsRect;
	Canvas->FillRect(rect, 0, 0, TCorners(), DEFAULT_OPACITY);

	if (test)
	{
	  g_debugDraw.Canvas = Canvas;
	  g_debugDraw.Canvas->Fill->Color = TAlphaColorRec::Yellow;
	  test->DrawTitle(entry->name);
	  g_debugDraw.Canvas->Fill->Color = TAlphaColorRec::Aqua;
	  test->Step(&settings);
	}
  }
  __finally
  {
	Canvas->EndScene();
  }
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::FormShow(TObject *Sender)
{
  ResetView();
  DrawLayoutChanged(0);
  TestChanged();
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::FormResize(TObject *Sender)
{
  ResetView();
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::Timer1Timer(TObject *Sender)
{
  if (test)
  {
    Invalidate();
  }
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::TestListChange(TObject *Sender)
{
  if (TestList->ItemIndex != testSelection)
  {
    testSelection = TestList->ItemIndex;
    TestChanged();
  }
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::FormKeyDown(TObject *Sender, WORD &Key, System::WideChar &KeyChar,
          TShiftState Shift)
{
  bool modCtrl = Shift.Contains(ssCtrl);
  switch(Key)
  {
  case vkEscape:
    // Quit
    Close();
    break;

  case vkLeft:
    // Pan left
    if (modCtrl && test)
    {
      b2Vec2 newOrigin(2.0f, 0.0f);
      test->ShiftOrigin(newOrigin);
    }
    else
    {
      g_camera.m_center.x -= 0.5f;
    }
    break;

  case vkRight:
    // Pan right
    if (modCtrl && test)
    {
      b2Vec2 newOrigin(-2.0f, 0.0f);
      test->ShiftOrigin(newOrigin);
    }
    else
    {
      g_camera.m_center.x += 0.5f;
    }
    break;

  case vkDown:
    // Pan down
    if (modCtrl && test)
    {
      b2Vec2 newOrigin(0.0f, 2.0f);
      test->ShiftOrigin(newOrigin);
    }
    else
    {
      g_camera.m_center.y -= 0.5f;
    }
    break;

  case vkUp:
    // Pan up
    if (modCtrl && test)
    {
      b2Vec2 newOrigin(0.0f, -2.0f);
      test->ShiftOrigin(newOrigin);
    }
    else
    {
      g_camera.m_center.y += 0.5f;
    }
    break;

  case vkHome:
    // Reset view
    g_camera.m_zoom = 1.0f;
    g_camera.m_center.Set(0.0f, 20.0f);
    break;

  case vkSpace:
    // Launch a bomb.
    if (test)
    {
      test->LaunchBomb();
    }
    break;

  default:
    switch (KeyChar)
    {
    case 'P':
    case 'p':
      // Pause
      settings.pause = !settings.pause;
      break;

    case ' ':
      // Launch a bomb.
      if (test)
      {
        test->LaunchBomb();
      }
      break;

    case 'R':
    case 'r':
      // Reset test
      delete test;
      test = entry->createFcn();
      break;

    default:
      if (test)
      {
        test->Keyboard(static_cast<int>(ToUpper(KeyChar)));
      }
    }
  }
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::FormKeyUp(TObject *Sender, WORD &Key, System::WideChar &KeyChar,
          TShiftState Shift)
{
  if (test)
  {
    test->KeyboardUp(static_cast<int>(ToUpper(KeyChar)));
  }
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::PaintBoxMouseDown(TObject *Sender, TMouseButton Button,
          TShiftState Shift, float X, float Y)
{
  if (Button == TMouseButton::mbLeft)
  {
    if (test)
    {
      b2Vec2 pt = g_debugDraw.ScreenToWorld(X, Y);
      if (Shift.Contains(ssShift))
        test->ShiftMouseDown(pt);
      else
        test->MouseDown(pt);
    }
  }
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::PaintBoxMouseMove(TObject *Sender, TShiftState Shift,
          float X, float Y)
{
  if (test)
  {
    b2Vec2 pt = g_debugDraw.ScreenToWorld(X, Y);
    test->MouseMove(pt);
  }
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::PaintBoxMouseUp(TObject *Sender, TMouseButton Button,
          TShiftState Shift, float X, float Y)
{
  if (test)
  {
    b2Vec2 pt = g_debugDraw.ScreenToWorld(X, Y);
    test->MouseUp(pt);
  }
}
//---------------------------------------------------------------------------


void __fastcall TTestBedForm::DrawLayoutChanged(TObject *Sender)
{
  settings.drawShapes = this->ShapesChk->IsChecked;
  settings.drawJoints = this->JointsChk->IsChecked;
  settings.drawAABBs = this->AABBsChk->IsChecked;
  settings.drawContactPoints = this->ContactPointsChk->IsChecked;
  settings.drawContactNormals = this->ContactNormalsChk->IsChecked;
  settings.drawContactImpulse = this->ContactImpulsesChk->IsChecked;
  settings.drawFrictionImpulse = this->FrictionImpulsesChk->IsChecked;
  settings.drawCOMs = this->CenterOfMassesChk->IsChecked;
  settings.drawStats = this->StatisticsChk->IsChecked;
  settings.drawProfile = this->ProfileChk->IsChecked;
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::PauseBtnClick(TObject *Sender)
{
  settings.pause = !settings.pause;
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::SingleStepBtnClick(TObject *Sender)
{
  settings.singleStep = !settings.singleStep;
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::RestartBtnClick(TObject *Sender)
{
  delete test;
  test = entry->createFcn();
}
//---------------------------------------------------------------------------

void __fastcall TTestBedForm::QuiBtnClick(TObject *Sender)
{
  Close();
}
//---------------------------------------------------------------------------


